NCMC Code Refactoring

nathanmlim opened the pull request:
This [WIP] is meant to refactor the NCMC core code. I've placed the new refactor code into a separate directory `blues_refactor` so that I could test them side-by-side to make sure I haven't introduced any new bugs during refactoring. Briefly, I've basically gone in and tried to reduce large code blocks in `ncmc.py` and moved them into their own functions. I've added 3 classes to handle some basic routines:
* `SimulationFactory` handles generating a set of 3 OpenMM Simulations objects for the MD, NCMC, Alchemical parts of the BLUES engine.
* `LigandModeller` intended to handle calculating any properties relating to the ligand/object we're moving around. As of now, we're just calculating particle masses but I think down the line we may want to calculate other properties on like protein residues or water densities etc.
* `ProposeMove` intended to be a class where we basically define our move sets.
Much of the simplifications from introducing these classes can be seen in the new example.py script

TODO:
* [x] Add Docstrings
* [x] Add unit tests for new classes
* [x] Finish up simplifying/moving the `ncmc.Simulation.run()` code blocks.

sgill2 commented:
So these are just my initial thoughts/questions on the code. 
1. I'm not too clear on the sims object that you use as the SImulation object argument. It looks like it's supposed to be an object that holds the openmm simulation, alch_sim and ncmc contexts. If that's the case what benefit is there doing it this way as  specifying them via separate arguments?
2. Could you elaborate on what the model parameter is supposed to be?

 Overall though, it seems more nicely structured than the previous code, and is heading in a good direction.

davidlmobley commented on README.md:
```diff
@@ -12,6 +12,7 @@ This also provides a prototype and validation of the SMIRFF SMIRKS-based force f
 * `run_scripts/` - example scripts to run blues
 * `systems/` - some example systems to run blues on.
 
+<<<<<<< HEAD
```
Do you have a merge conflict here?

nathanmlim replied:
Oops. Will fix that.

davidlmobley commented on blues/example.py:
```diff
@@ -1,5 +1,14 @@
+
+* Benchmarking related code adapted from:
+https://github.com/pandegroup/openmm/blob/master/examples/benchmark.py
+(Authors: Peter Eastmen)
```
Eastman

davidlmobley commented on blues/example.py:
```diff
@@ -16,7 +25,11 @@
 def runNCMC(platform_name):
+=======
+def runNCMC(options):
+>>>>>>> 63cd22390b2e6595bd1defa81aa758e62ae31fc2:examples/example.py
```
Merge conflicts?

davidlmobley commented on README.md:
````diff
@@ -26,6 +27,33 @@ Install from source
+Install the required packages into a new environment
+
+```
+conda create -c omnia -c omnia/label/dev -n blues python=3.5 openmm==7.0.1 openmmtools alchemy mdtraj
````
Wait, are we in omnia now?

And, openmm 7.0.1 not 7.1?

davidlmobley replied:
(Oh, I see, this is creating a blues environment. Sorry, my bad. But the question on OpenMM still stands.)

nathanmlim replied:
It seems I forgot to commit the conflicts in the README.md I'll fix that.

davidlmobley commented on blues_refactor/example.py:
```diff
@@ -0,0 +1,85 @@
+            'verbose' : True }
+
+    #Defines ncmc move eqns for lambda peturbation of sterics/electrostatics
+    opt['functions'] = { 'lambda_sterics' : 'step(0.199999-lambda) + step(lambda-0.2)*step(0.8-lambda)*abs(lambda-0.5)*1/0.3 + step(lambda-0.800001)',
```
What does this mean? (OK, I can work through it, but it's a bit opaque). What does it control?

davidlmobley replied:
(Obviously, it controls lambda_sterics... But what exactly is going on here? )

sgill2 replied:
I'm still in the process of forming all my opinions, so I'll keep writing them down as I go. 
I think passing an opt dictionary for run() that handles  runoptions makes things unclear to the user what actually can be done with it. I'd much rather have too many parameters, which users at least know exist, than having a mysterious, but important parameter.

davidlmobley replied:
@sgill2 - having a dictionary of options can be helpful as it can allow for extension/modification of the options without modifying a list of required arguments. I'm less clear as to whether this specific implementation is the best way to do it or not, but that's the rationale I believe (and to a large extent I buy into it).

@nathanmlim - further comments?

nathanmlim replied:
All I've done here is take this [line](https://github.com/MobleyLab/blues/blob/master/blues/example.py#L29)  and chucked it into the `opt` dictionary. In retrospect, I should move this into the `SimulationFactory` as it only appears to be needed during the nc_integrator creation.

davidlmobley commented on blues_refactor/example.py:
```diff
@@ -0,0 +1,85 @@
+    sims.createSimulationSet(opt)
+
+    # Calculate particle masses of object to be moved
+    from blues.modeller import LigandModeller
```
I wonder if "LigandModeller" is really the best name for this thing if you're expecting perhaps generalizing to calculation of other properties later. What if we wanted to apply BLUES to a protein sidechain (rotamer sampling) or to swapping a ligand with water? (Or to jumping a water, rather than a ligand, between sites?)

nathanmlim replied:
I'm not really attached to naming and couldn't think of a good name on the fly. So if you have suggestions I'm open. I don't even think `Modeller` was a good choice either since OpenMM uses this to manipulate objects rather than get properties on them.

davidlmobley replied:
@sgill2 - suggestions? What is the general purpose of this thing? `ObjectProperties`? `MoveProperties`? `MoveObject`? `MovingObject`? The last one seems to appeal to me most, as it might or might not be a ligand, and I also don't like the `modeller` terminology because of (a) UCSF Modeller, and (b) `Modeller` within OpenMM, as you note, @nathanmlim .

nathanmlim replied:
I think `ObjectProperties` seems like the best bet here.

davidlmobley commented on blues_refactor/example.py:
```diff
@@ -0,0 +1,85 @@
+    model.calculateCOM()
+
+    # Propse some move
+    rot_move = ncmc.ProposeMove(sims.nc, model)
```
Should `ProposeMove` be `MoveProposal` or similar? Right now it reads like a function name (action implied!) to me, so I was confused when I read it -- I thought "wait, you can't be proposing a move yet, as you haven't said what kind of move you want."

nathanmlim replied:
Okay, I can change this to `MoveProposal`

davidlmobley commented on blues_refactor/example.py:
```diff
@@ -0,0 +1,85 @@
+    # Propse some move
+    rot_move = ncmc.ProposeMove(sims.nc, model)
+    rot_step = (opt['nstepsNC']  / 2) - 1
+    nc_move = { 'type' : 'rotation',
```
@sgill2 - should we be more specific than "rotation", do you think? Since we might imagine other types of rotations (e.g. rotation matrix rather than random quaternion) maybe this should be "random_rotation"?

sgill2 replied:
More specific is probably better, especially since we don't know what other kinds of moves we're adding; more specificity can help avoid name overlaps, so "random_rotation", would be better.

nathanmlim replied:
Works for me.

davidlmobley commented on blues_refactor/ncmc.py:
```diff
@@ -0,0 +1,216 @@
+
+        if log_ncmc > randnum:
+            self.accept += 1
+            print('NCMC MOVE ACCETPED: {} > {}'.format(log_ncmc, randnum) )
```
"ACCEPTED"

davidlmobley commented on blues_refactor/ncmc.py:
```diff
@@ -0,0 +1,216 @@
+        randnum =  math.log(np.random.random())
+
+        ### Compute Alchemical Correction Term
+        if not np.isnan(log_ncmc):
```
@sgill2 - what happens if it IS a NAN? Not totally obvious to me it doesn't do something strange (does it accept since `log_ncmc > randnum` below?

sgill2 replied:
I seem to remember something weird happening with NANs in the past, but trying some logical operations with NANs doesn't seem to suggest they would be accepted. I'll make a separate issue to see what happens with the NAN checking removed.

nathanmlim replied:
This originally had the `if alchemical_correction==True` statement with checking if `log_ncmc` is NaN. So I _think_ that if you were to add the alchemical correction term, it won't let you add it to NaN. Since I've removed the alchemical_correction switch, should we turn NaN into 0 so that we can still add the alchemical correction term?

davidlmobley commented on blues_refactor/tests/test_mass.py:
```diff
@@ -0,0 +1,57 @@
+        npos[i] = rotated_particles[index]
+    r_com_coord = test_class.calculate_com(npos)
+    for i in range(3):
+        np.testing.assert_almost_equal(np.asarray(r_com_coord._value)[i], np.asarray(com_coord._value)[i], decimal=4)
```
Odd character on this line? (Or missing newline?)

nathanmlim commented:
> So these are just my initial thoughts/questions on the code.
> 
> I'm not too clear on the sims object that you use as the SImulation object argument. It looks like it's supposed to be an object that holds the openmm simulation, alch_sim and ncmc contexts. If that's the case what benefit is there doing it this way as specifying them via separate arguments?
> Could you elaborate on what the model parameter is supposed to be?
> Overall though, it seems more nicely structured than the previous code, and is heading in a good direction.

The `sims` object is just a dictionary containing the 3 simulations (md, alch, ncmc) that is generated out of the `SimulationFactory` class. Really you can name it whatever you want since it a `SimulationFactory` object. We have to lug these around together anyways and pass it into the BLUES engine itself to do the run, so for simplicity I just put them into a dictionary to pass as a single object.

nathanmlim commented:
Walked through this with @sgill2 and we have reviewed the refactored code and are certain it does the some process. We did identify some potential bugs and made some other changes that may require another look @davidlmobley. Changes are noted below:

- Changed the name from `ModelProperties` to `Model`
- Changed the lambda functions from [this](https://github.com/MobleyLab/blues/blob/master/blues/example.py#L29) to 
```
functions = { 'lambda_sterics' : 'min(1, (1/0.3)*abs(lambda-0.5))',
 'lambda_electrostatics' : 'step(0.2-lambda) - 1/0.2*lambda*step(0.2-lambda) + 1/0.2*(lambda-0.8)*step(lambda-0.8)' }
```
- Added the negative sign back to the alchemical correction factor in [this line](https://github.com/MobleyLab/blues/blob/master/blues/ncmc.py#L518) 

Note: Since the `smartdart.py` has not been refactored into the new framework and inherits from the old `ncmc.py` it's tests have been disabled.

davidlmobley commented:
@nathanmlim - it would be very helpful to have a comment before the `functions` line relating to lambda which explains a little more about what these `lambda_sterics`, `lambda_electrostatics`, etc. are doing and how they are being written. The factors `1/0.3` and `1/0.2` are very non-obvious to me, as are the use of `step` (which is not defined here); understanding this would presumably require reading the documentation of the relevant code, but this is not linked to. So I'd have to do some archaeology to figure out what this means.

davidlmobley commented on README.md:
```diff
@@ -41,9 +41,6 @@ This package takes advantage of non-candidate equilibrium monte carlo moves (NCM
 The heart of the package is found in `blues/ncmc_switching.py`. This holds the framework for NCMC. Particularly, it contains the integrator class that calculates the work done during the NCMC move. It also controls the lambda scaling of parameters. Currently the alchemy package is used to generate the lambda parameters for the ligand, which can potentially modify 5 parameters–sterics, electrostat ...
 The class SimNCMC in `blues/ncmc.py` serves as a wrapper for running ncmc simulations. SimNCMC.runSim() actually runs the simulation.
 
-## Example Use
```
Wait, we have no example anymore? What are the plans on this?

davidlmobley replied:
Oh, wait, I still see `blues/example.py` here -- why no mention in README now?

nathanmlim replied:
I'm not sure why the entire block got removed. It might have gotten accidently deleted when I was trying to resolve conflicts. It should have just been changed to reference the `blues/example.py` location.

nathanmlim commented:
@davidlmobley: I think that would be a better question for @sgill2 since I just copied over what he had suggested.

davidlmobley commented:
I agree with the sign of the alchemical correction factor, though I've challenged @sgill2 to think about how to check it. But it seems correct to me.

davidlmobley commented on blues/example.py:
```diff
@@ -1,117 +1,65 @@
+    simulations.createSimulationSet()
+
+    blues = Simulation(simulations, ligand, ligand_mover, **opt)
+    blues.run()
```
The example sure becomes nice and concise now. Very nice. :)

davidlmobley commented on blues/ncmc.py:
```diff
@@ -1,580 +1,332 @@
+    ligand.masses : list of particle masses of the ligand with units.
+    ligand.totalmass : integer of the total mass of the ligand.
+
+    #Dynamic attribtues that must be updated with each iteration
```
"attributes"

davidlmobley commented on blues/ncmc.py:
```diff
@@ -1,580 +1,332 @@
+        """
+        if ncmc:
+            #Defines ncmc move eqns for lambda peturbation of sterics/electrostatics
+            functions = { 'lambda_sterics' : 'min(1, (1/0.3)*abs(lambda-0.5))',
```
As noted, it still bothers me I can't understand this with the info in the comments.

davidlmobley commented on blues/ncmc.py:
```diff
@@ -1,580 +1,332 @@
+        self.md = self.generateSimFromStruct(self.structure, self.system, **self.opt)
+        self.alch = self.generateSimFromStruct(self.structure, self.system,  **self.opt)
+        self.nc = self.generateSimFromStruct(self.structure, self.alch_system,
+                                            ncmc=True, **self.opt)
```
Odd character at end of line 332?

davidlmobley commented on blues/ncmc_switching.py:
```diff
@@ -1,5 +1,14 @@
+*Adapted from source code:
 #https://github.com/choderalab/perses/blob/master/perses/annihilation/ncmc_switching.py
+
+Authors: Patrick B. Grinaway, Julie M. Behr, and John D. Chodera
```
@sgill2 - add note on outline of paper to see if Julie should be an author.

davidlmobley commented on blues/simulation.py:
```diff
@@ -0,0 +1,284 @@
+
+    def getStateInfo(self, context, parameters):
+        """Function that gets the State information from the given context and
+        list of parameters to queuey it with.
```
query

davidlmobley commented on blues/simulation.py:
```diff
@@ -0,0 +1,284 @@
+        if log_ncmc > randnum:
+            self.accept += 1
+            print('NCMC MOVE ACCEPTED: log_ncmc {} > randnum {}'.format(log_ncmc, randnum) )
+            #print('accCounter', float(self.accept)/float(stepsdone+1), self.accept)
```
Remove old comments? Otherwise use a debug/logging option so they can be used if needed?

davidlmobley commented on blues/simulation.py:
```diff
@@ -0,0 +1,284 @@
+        self.nc_context.setVelocities(md_state0['velocities'])
+
+    def run(self):
+        """Function that runs the BLUES engine that iterates of the actions:
```
"that iterates of the actions" -> "to iterate over the actions"...?

nathanmlim commented:
@davidlmobley  
> Do you want to include some of your info on code architecture/organization in a "developer documentation" README or something similar? (Note you can have additional .md files in the main directory which could be linked from the main README.md). I'm just thinking, @nathanmlim , that this could be helpful for people who are getting involved with developing so you don't always have to show them your slides/explain.

This is a good idea, I'll make the diagram a bit more detailed so it can serve as a quick overview image of the engine.

davidlmobley commented:
@sgill2 - can you get Nathan info to include on how the lambda vectors work? either an explanation or a documentation link to include in the comments or something similar?

sgill2 commented:
@davidlmobley Could you clarify what you mean by the lambda vector? Do you mean the equations for the functions that control the lambda sterics and electrostatics ?

davidlmobley commented:
> @davidlmobley Could you clarify what you mean by the lambda vector? Do you mean the equations for the functions that control the lambda sterics and electrostatics ?

Yes, @sgill2 - see my comments above to Nathan about that. Basically from the code/comments in the code it's impossible for me to know what the lambda sterics/lambda electrostatics stuff means. I'd like this to be better explained in comments, and/or there to be links to documentation. I don't get the factor 1/3, or what `step` does, etc. Presumably I could figure it out with some archeology/Googling but since it's an obvious question it makes sense to have SOME info in comments or in the code.

davidlmobley commented:
Good to merge once tests pass/you're done, I think, @nathanmlim .

nathanmlim commented:
Added in some comments from Sam for the lambda functions. Also added in  `devdocs/` folder which has a somewhat incomplete class diagram (haven't gone in and added in the full list of attributes to the other classes) but I'll do that in a separate PR so we can get this merged.

nathanmlim commented:
Merged.

davidlmobley commented:
Yay, congrats!
